File: go/gts/compact/compact_test.go
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: Apache-2.0

package compact

import (
	"bytes"
	"os"
	"path/filepath"
	"testing"

	"go.blackcatinformatics.ca/gts/model"
	"go.blackcatinformatics.ca/gts/reader"
	"go.blackcatinformatics.ca/gts/wire"
	"go.blackcatinformatics.ca/gts/writer"
)

func vector(t *testing.T, name string) []byte {
	t.Helper()
	dir, err := filepath.Abs("../../../generated/gts-vectors")
	if err != nil {
		t.Fatal(err)
	}
	//nolint:gosec // test reads a frozen conformance vector by caller-supplied name.
	data, err := os.ReadFile(filepath.Join(dir, name))
	if err != nil {
		t.Fatal(err)
	}
	return data
}

// TestCompactReproducesFrozenVector is the cross-engine determinism oracle
// (§14.1): compacting the frozen vector 25 bytes with the frozen timestamp
// MUST reproduce the frozen vector 25b bytes exactly.
func TestCompactReproducesFrozenVector(t *testing.T) {
	source := vector(t, "25-streamable-source.gts")
	expected := vector(t, "25b-streamable-compacted.gts")
	got, err := Streamable(source, "2026-01-01T00:00:00Z", false)
	if err != nil {
		t.Fatalf("compact refused: %v", err)
	}
	if !bytes.Equal(got, expected) {
		t.Fatalf("compacted bytes diverge from the frozen oracle: got %d bytes, want %d bytes", len(got), len(expected))
	}
}

func TestCompactOutputVerifiesStreamable(t *testing.T) {
	source := vector(t, "25-streamable-source.gts")
	got, err := Streamable(source, "2026-01-01T00:00:00Z", false)
	if err != nil {
		t.Fatalf("compact refused: %v", err)
	}
	g := reader.Read(got, true, nil)
	if len(g.Diagnostics) > 0 {
		t.Fatalf("compacted output has diagnostics: %v", g.Diagnostics)
	}
	if len(g.SegmentStreamable) != 1 {
		t.Fatalf("expected one segment, got %d", len(g.SegmentStreamable))
	}
	info := g.SegmentStreamable[0]
	if !info.Claimed || info.Tail != 0 || info.Covered == 0 {
		t.Fatalf("unexpected layout state: %+v", info)
	}
}

func TestCompactSealOriginalCarriesSourceBlob(t *testing.T) {
	source := vector(t, "25-streamable-source.gts")
	got, err := Streamable(source, "2026-01-01T00:00:00Z", true)
	if err != nil {
		t.Fatalf("compact refused: %v", err)
	}
	g := reader.Read(got, true, nil)
	if len(g.Diagnostics) > 0 {
		t.Fatalf("sealed output has diagnostics: %v", g.Diagnostics)
	}
	sealed := wire.DigestStr(source)
	data := blobBytes(g, sealed)
	if !bytes.Equal(data, source) {
		t.Fatalf("sealed blob does not carry the verbatim source bytes")
	}
	if mt, _ := blobMetaString(g, sealed, "mt"); mt != "application/gts" {
		t.Fatalf("sealed blob media type: %q", mt)
	}
	if rep, _ := blobMetaString(g, sealed, "rep"); rep != "source" {
		t.Fatalf("sealed blob rep: %q", rep)
	}
}

func TestCompactRefusesEvidenceWithoutSeal(t *testing.T) {
	w := writer.New("evidence")
	w.AddTerms([]model.Term{
		{Kind: model.Iri, Value: "https://example.org/Cat"},
		{Kind: model.Iri, Value: "http://www.w3.org/2000/01/rdf-schema#label"},
		{Kind: model.Literal, Value: "Cat", Lang: "en"},
	})
	w.AddQuads([]model.Quad{{S: 0, P: 1, O: 2}})
	data := w.ToBytes()

	if _, err := Streamable(data, "2026-01-01T00:00:00Z", false); err == nil {
		t.Fatal("expected a refusal for evidence input without seal-original")
	}
	if _, err := Streamable(data, "2026-01-01T00:00:00Z", true); err != nil {
		t.Fatalf("seal-original should permit evidence compaction: %v", err)
	}
}

func TestCompactRefusesFrameSuppression(t *testing.T) {
	w := writer.New("generic")
	w.AddTerms([]model.Term{
		{Kind: model.Iri, Value: "https://example.org/Cat"},
		{Kind: model.Iri, Value: "http://www.w3.org/2000/01/rdf-schema#label"},
		{Kind: model.Literal, Value: "Cat", Lang: "en"},
	})
	w.AddQuads([]model.Quad{{S: 0, P: 1, O: 2}})
	w.AddSuppress([]interface{}{
		map[interface{}]interface{}{"kind": "frame", "digest": "blake3:00"},
	}, "", nil)
	data := w.ToBytes()

	_, err := Streamable(data, "2026-01-01T00:00:00Z", false)
	if err == nil {
		t.Fatal("expected a refusal for a frame-addressed suppression")
	}
	if _, ok := err.(*RefusedError); !ok {
		t.Fatalf("expected *RefusedError, got %T", err)
	}
}

File: ts/gts/tests/compact.test.ts
// SPDX-FileCopyrightText: 2026 Blackcat Informatics® Inc. <paudley@blackcatinformatics.ca>
// SPDX-License-Identifier: Apache-2.0

import { readFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { test } from "node:test";
import assert from "node:assert/strict";
import { compactStreamable, CompactRefusedError } from "../src/compact.js";
import { Read } from "../src/reader.js";
import { Writer } from "../src/writer.js";
import { TermKind } from "../src/model.js";
import { hex, digestStr } from "../src/wire.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(__dirname, "../../../../");
const vectorsDir = join(repoRoot, "generated", "gts-vectors");

test("compact reproduces the frozen 25b bytes exactly (§14.1 determinism)", () => {
    const source = new Uint8Array(
        readFileSync(join(vectorsDir, "25-streamable-source.gts")),
    );
    const frozen = new Uint8Array(
        readFileSync(join(vectorsDir, "25b-streamable-compacted.gts")),
    );
    const out = compactStreamable(source, {
        timestamp: "2026-01-01T00:00:00Z",
    });
    assert.equal(hex(out), hex(frozen), "compacted bytes differ from frozen");
});

test("compact output claims streamable with zero tail and verifies clean", () => {
    const source = new Uint8Array(
        readFileSync(join(vectorsDir, "25-streamable-source.gts")),
    );
    const out = compactStreamable(source, {
        timestamp: "2026-01-01T00:00:00Z",
    });
    const g = Read(out, true);
    assert.equal(g.diagnostics.length, 0);
    assert.equal(g.segmentStreamable.length, 1);
    assert.ok(g.segmentStreamable[0].claimed);
    assert.equal(g.segmentStreamable[0].tail, 0);
});

test("compact carries the reasoned suppression forward, ids shifted (§10.1)", () => {
    const source = new Uint8Array(
        readFileSync(join(vectorsDir, "25-streamable-source.gts")),
    );
    const src = Read(source, true);
    const out = Read(
        compactStreamable(source, { timestamp: "2026-01-01T00:00:00Z" }),
        true,
    );
    assert.equal(out.suppressions.length, src.suppressions.length);
    assert.equal(out.suppressions[0].reason, "superseded");
    const target = out.suppressions[0].targets[0];
    assert.ok(target instanceof Map);
    assert.equal(target.get("kind"), "term");
    const tid = target.get("id") as number;
    // The shifted target resolves to the same term value as in the source.
    const srcTarget = src.suppressions[0].targets[0] as Map<unknown, unknown>;
    const srcTid = srcTarget.get("id") as number;
    assert.equal(out.terms[tid].value, src.terms[srcTid].value);
});

test("compact refuses an evidence input without --seal-original (§10.1)", () => {
    const w = new Writer("evidence");
    w.addTerms([
        { kind: TermKind.Iri, value: "https://example.org/Cat" },
        {
            kind: TermKind.Iri,
            value: "http://www.w3.org/2000/01/rdf-schema#label",
        },
        { kind: TermKind.Literal, value: "Cat", lang: "en" },
    ]);
    w.addQuads([{ s: 0, p: 1, o: 2 }]);
    const data = w.toBytes();
    assert.throws(
        () => compactStreamable(data, { timestamp: "2026-01-01T00:00:00Z" }),
        CompactRefusedError,
    );
    const sealed = compactStreamable(data, {
        timestamp: "2026-01-01T00:00:00Z",
        sealOriginal: true,
    });
    const g = Read(sealed, true);
    assert.equal(g.diagnostics.length, 0);
    // The sealed original travels as a blob, role "source", byte-intact.
    const sealedDigest = digestStr(data);
    const blob = g.blobs.find((b) => b.digest === sealedDigest);
    assert.ok(blob, "sealed original blob missing");
    assert.equal(hex(blob.data), hex(data));
});

test("compact is reproducible for a fixed timestamp", () => {
    const source = new Uint8Array(
        readFileSync(join(vectorsDir, "25-streamable-source.gts")),
    );
    const a = compactStreamable(source, { timestamp: "2026-01-01T00:00:00Z" });
    const b = compactStreamable(source, { timestamp: "2026-01-01T00:00:00Z" });
    assert.equal(hex(a), hex(b));
});
